# # Explanation about `ggdb_enzymes.json`
#
# ## Format
#
# - "name": gene symbol of the enzyme
# - "rules": a list of rules, each rule is a dictionary with the following keys:
#     - "acceptor": the motif recognized by the enzyme
#     - "acceptor_alignment": alignment of the acceptor motif
#     - "product": the product generated by the enzyme,
#       should be exactly one residue longer than the acceptor
# - "rejects": a list of exceptions to "acceptor",
#   each exception is a dictionary with the following keys:
#     - "motif": the motif to be rejected
#     - "alignment": alignment of the motif
# - "markers": a list of motifs that indicate the enzyme is involved in the biosynthesis of a glycan,
#   each marker is a dictionary with the following keys:
#     - "motif": the marker motif
#     - "alignment": alignment of the marker
# - "type": "GT" for glycosyltransferase or "GD" for exoglycosidase
# - "species": species of the enzyme
#
# ## Additional Notes
#
# ### Rules about the ALG family
#
# These enzymes involve in the synthesis of the dolichol-linked precursor.
# The oligosaccharide part is linked to P-P-Dolichol throught a-linkage.
# However, when the whole glycan is transfered to Asn, the linkage becomes b-linkage.
# Therefore, although these enzymes actually recognize motifs with a-reducing-ends,
# we use "b" as the reducing-end in the acceptor and product.
#
# ### ALG markers
#
# All N-glycans observed on Asn must have been processed by all ALG family.
# So the "marker" field for these enzymes is just the N-glycan core.
